lib/session.ts

```diff
@@ -0,0 +1,136 @@
+    try {
+        const { payload } = await jwtVerify(session, encodedKey, {
+            algorithms: ['HS256'],
+        });
+        return payload;
+    } catch (error) {
+        console.log('Failed to verify session:', session);
+    }
+}
+
+export async function getSessionUser() {
+    "use server";
+    // TODO refresh
+    const session = cookies().get('session');
+    if (session?.value) {
+        const sessionPayload = await decrypt(session.value);
+        if (sessionPayload?.userId) {
+            const email: string = sessionPayload.userId as string;
+            const personalOAuthClient = await getPersonalOAuthClient(email);
+            if (personalOAuthClient !== undefined) {
+                experimental_taintObjectReference("No leaky pls", personalOAuthClient);
+                // Store refreshed Google tokens
+                personalOAuthClient.on("tokens", (tokens => {
+                    console.log("Refreshed Google tokens for", email);
+                    setPersonalOAuthClient(email, tokens);
+                }));
+            }
+            return personalOAuthClient;
+        } else {
+            return undefined;
+        }
+    }
+}
+
+export async function getSessionName(): Promise<string | undefined> {
+    "use server";
+    const session = cookies().get('session');
+    if (session?.value) {
+        const sessionPayload = await decrypt(session.value);
+        return sessionPayload?.userName as (string | undefined);
+    }
+    return undefined;
+}
+
+export async function hasSessionUser(): Promise<boolean> {
+    "use server";
+    try {
+        const session = cookies().get('session');
+        if (session?.value) {
+            const sessionPayload = await decrypt(session.value);
+            if (sessionPayload?.userId) {
+                const email: string = sessionPayload.userId as string;
+                const personalOAuthClient = await getPersonalOAuthClient(email);
+
+                if (personalOAuthClient) {
+                    experimental_taintObjectReference("No leaky pls", personalOAuthClient);
+                    return true;
+                }
+            }
+        }
+        return false;
+    } catch (err) {
+        return false;
+    }
+}
+
+export async function createSession(userId: string, userName: string) {
+    const expiresAt = dayjs().add(7, "days").toDate();
+    const session = await encrypt({ userId, userName, expiresAt });
+
+    cookies().set('session', session, {
+        httpOnly: true,
+        secure: true,
+        expires: expiresAt,
+        sameSite: 'lax',
+        path: '/',
+    });
+}
+
+export async function deleteSession(revokeCredentials = true) {
+    const session = cookies().get('session');
+    if (session?.value) {
+        const sessionPayload = await decrypt(session.value).catch();
+        if (sessionPayload?.userId) {
+            const email: string = sessionPayload.userId as string;
+            console.log("Logout from", email);
```

Comment:
Zullen we geen console logjes op main zetten?

Reply:
Deze zijn er expres ingelaten om eventuele bugs makkelijker te kunnen onderzoeken. Ik zal ze voor nu verwijderen.